pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandlerTest.java

```diff
@@ -0,0 +1,374 @@
+    assertEquals(operationMap, Collections.EMPTY_MAP);
+
+    // TEST4: Add random index
+    indexLoadingConfig = new IndexLoadingConfig(null, _tableConfig);
+    indexLoadingConfig.getTextIndexColumns().add(DICT_INTEGER);
+    indexLoadingConfig.getTextIndexColumns().add(LZ4_INTEGER);
+    fwdIndexHandler = new ForwardIndexHandler(_existingSegmentMetadata, indexLoadingConfig);
+    operationMap = fwdIndexHandler.computeOperation(_writer);
+    assertEquals(operationMap, Collections.EMPTY_MAP);
+
+    // TEST5: Change compression
+
+    // Create new tableConfig with the modified fieldConfigs.
+    List<FieldConfig> fieldConfigs = new ArrayList<>(_tableConfig.getFieldConfigList());
+    FieldConfig config = fieldConfigs.remove(0);
+    FieldConfig newConfig = new FieldConfig(config.getName(), FieldConfig.EncodingType.RAW, Collections.emptyList(),
+        FieldConfig.CompressionCodec.ZSTANDARD, null);
+    fieldConfigs.add(newConfig);
+    TableConfig tableConfig =
+        new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME).setNoDictionaryColumns(_noDictionaryColumns)
+            .setFieldConfigList(fieldConfigs).build();
+    tableConfig.setFieldConfigList(fieldConfigs);
+
+    indexLoadingConfig = new IndexLoadingConfig(null, tableConfig);
+    fwdIndexHandler = new ForwardIndexHandler(_existingSegmentMetadata, indexLoadingConfig);
+    operationMap = fwdIndexHandler.computeOperation(_writer);
+    assertEquals(operationMap.size(), 1);
+    assertEquals(operationMap.get(config.getName()), ForwardIndexHandler.Operation.CHANGE_RAW_COMPRESSION_TYPE);
+
+    // TEST6: Change compression and add index
+    fieldConfigs = new ArrayList<>(_tableConfig.getFieldConfigList());
+    config = fieldConfigs.remove(0);
+    newConfig = new FieldConfig(config.getName(), FieldConfig.EncodingType.RAW, Collections.emptyList(),
+        FieldConfig.CompressionCodec.ZSTANDARD, null);
+    fieldConfigs.add(newConfig);
+    tableConfig =
+        new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME).setNoDictionaryColumns(_noDictionaryColumns)
+            .setFieldConfigList(fieldConfigs).build();
+    tableConfig.setFieldConfigList(fieldConfigs);
+
+    indexLoadingConfig = new IndexLoadingConfig(null, tableConfig);
+    indexLoadingConfig.getTextIndexColumns().add(config.getName());
+    indexLoadingConfig.getInvertedIndexColumns().add(config.getName());
+    fwdIndexHandler = new ForwardIndexHandler(_existingSegmentMetadata, indexLoadingConfig);
+    operationMap = fwdIndexHandler.computeOperation(_writer);
+    assertEquals(operationMap.size(), 1);
+    assertEquals(operationMap.get(config.getName()), ForwardIndexHandler.Operation.CHANGE_RAW_COMPRESSION_TYPE);
+  }
+
+  @Test
+  public void testRewriteRawForwardIndex()
+      throws Exception {
+    List<String> allCompressionTypes = Arrays.asList("PASS_THROUGH", "ZSTANDARD", "LZ4", "SNAPPY");
+
+    for (int i = 0; i < _noDictionaryColumns.size(); i++) {
```

Comment:
Can we add test for bulk / batch changes as well ?

Reply:
Done.